Analysis/PIDTools/src/Compute_dEdxProcessor.cc

```diff
@@ -59,28 +89,71 @@ Compute_dEdxProcessor::Compute_dEdxProcessor()
-			  std::string("MarlinTrkTracks"));
+  registerProcessorParameter( "dxStrategy",
+			      "ID of the track length calculation: 1 - hit-to-hit distance; 2 - hit-to-hit path length of projected hits; 3 - path over hit row",
+			      _dxStrategy,
+			      int(1));
+
+  registerProcessorParameter( "StrategyCompHist",
+			      "If set, an AIDA root file will be generated with Bethe-Bloch histograms (dE/dx over momentum) for each dx strategy (default: false).",
+			      _StratCompHist,
+			      bool(false));
+
+  registerProcessorParameter( "StrategyCompHistWeight",
+			      "If set, the Bethe-Bloch histograms will have a sqrt(nHits) weighting (default: false).",
+			      _StratCompHistWeight,
+			      bool(false));
+
+  registerProcessorParameter( "StrategyCompHistFiles",
+			      "File names of the generated dx strategy comparison histograms (if chosen). The respective strategy number and '.png' is added.",
+			      _StratCompHistFiles,
+			      std::string("dEdx_Histo_Strategy"));
+
+
 } 
 
 void Compute_dEdxProcessor::init() { 
   streamlog_out(DEBUG) << "   init called  " << std::endl ;
   
-  //_mydEdx = new ComputeddEdx(); 
+  // it's usually a good idea to
+  printParameters();
   
-  //get TPC inner radius
+  //get TPC radii, pad height and B field
   dd4hep::Detector& lcdd = dd4hep::Detector::getInstance();
   dd4hep::DetElement tpcDE = lcdd.detector("TPC") ;
   const dd4hep::rec::FixedPadSizeTPCData* tpc = tpcDE.extension<dd4hep::rec::FixedPadSizeTPCData>() ;
-  _TPC_inner = tpc->rMinReadout ;
+  _TPC_inner = tpc->rMinReadout/dd4hep::mm;
+  _TPC_outer = tpc->rMaxReadout/dd4hep::mm;
+  _TPC_padHeight = tpc->padHeight/dd4hep::mm;
+  double bfieldV[3] ;
+  lcdd.field().magneticField( { 0., 0., 0. }  , bfieldV  ) ;
+  _bField = bfieldV[2]/dd4hep::tesla ;
 
   engine = new std::default_random_engine();
 
-  printParameters();
-  
+  streamlog_out(MESSAGE) << "TPC inner radius = " << _TPC_inner << " mm   TPC outer radius = " << _TPC_outer << " mm"<< std::endl;
+  streamlog_out(MESSAGE) << "TPC pad height = " << _TPC_padHeight << " mm   B field = " << _bField << " T" <<std::endl;
+  streamlog_out(DEBUG) << "dd4hep::mm = " << dd4hep::mm << "   dd4hep::cm = " << dd4hep::cm << std::endl;
+
+  if (_StratCompHist)
+  {
+    // prepare the histogram bins - this could also be implemented as processor parameters
+    const int nBinsX=1000, nBinsY=300;
+    double minBinX=-1, maxBinX=2;  // log10(min/max momentum / GeV)
+    double minBinY= 0, maxBinY=1e-6;
+    double histbinsX[nBinsX+1], histbinsY[nBinsY+1];
+    for (int i=0; i<nBinsX+1; i++) histbinsX[i] = pow( 10, minBinX + (maxBinX-minBinX)*i/(float)(nBinsX) );
+    for (int i=0; i<nBinsY+1; i++) histbinsY[i] = minBinY + (maxBinY-minBinY)*i/(float)(nBinsY);
+
+    // create the histograms
+    AIDAProcessor::tree(this);
+    _BBHist_Strategy1 = new TH2D("BBHist_Strategy1","Bethe-Bloch curve for dx strategy 1: hit-to-hit distance",nBinsX,histbinsX,nBinsY,histbinsY);
```

Comment:
`AIDAProcessor::tree(this)` does not assume you are using a ROOT implementation. You will get memory leak in this case.
If you want to use the AIDA processor, then use the [AIDA histogram factory](https://github.com/iLCSoft/Marlin/blob/master/source/include/marlin/AIDAProcessor.h#L84) : 
```cpp
auto factory = AIDAProcessor::histogramFactory(this);
auto histo = factory->createHistogram2D("BBHist_Strategy1",  "Bethe-Bloch curve for dx strategy 1: hit-to-hit distance", nBinsX, minBinX, maxBinX, nBinsY, minBinY, maxBinY);
// example of filling later in the code
histo->fill(mom, dedx);
```
Your second solution is to write your histogram in a separate root file. Then you need an additional processor parameter for the file name. In this case you can also write your TImage in this TFile instead of a separate png file.

Reply:
@rete technically you are of course right - but RAIDA is the only AIDA implementation that we ever used. So I think it is fine to instantiate ROOT histos after a call to `AIDAProcessor::tree(this);` Then every ROOT object goes to the directory with the processor name in the main (RAIDA)ROOT file ...

Reply:
Ok, so If I can leave it like that, then I'm done with the changes and would attempt a new pull request - permission granted?

Reply:
You don't have to make a new pull request, just push to your `improveDeDx` branch, it will update the pull request.